libs/core/libs/shared/src/document_repo.rs

```diff
@@ -1,15 +1,156 @@
 use crate::core_config::Config;
-use crate::crypto::*;
 use crate::file_metadata::DocumentHmac;
-use crate::{SharedErrorKind, SharedResult};
-use std::collections::HashSet;
+use crate::SharedResult;
+use crate::{crypto::*, SharedErrorKind};
+use serde::{Deserialize, Serialize};
+use std::cmp::Ordering;
+use std::collections::{HashMap, HashSet};
 use std::convert::TryInto;
 use std::fs::{self, File, OpenOptions};
+use std::hash::Hash;
 use std::io::{ErrorKind, Read, Write};
 use std::path::Path;
 use tracing::*;
 use uuid::Uuid;
 
+#[derive(Debug, Clone, Serialize, Deserialize, Ord, PartialEq, PartialOrd, Eq, Hash)]
+pub enum DocEvent {
+    Read(Uuid, i64),
+    Write(Uuid, i64),
+}
+impl DocEvent {
+    pub fn timestamp(&self) -> i64 {
+        match *self {
+            DocEvent::Read(_, timestamp) => timestamp,
+            DocEvent::Write(_, timestamp) => timestamp,
+        }
+    }
+    pub fn id(&self) -> Uuid {
+        match *self {
+            DocEvent::Read(id, _) => id,
+            DocEvent::Write(id, _) => id,
+        }
+    }
+}
+
+#[derive(Default, Copy, Clone, PartialEq)]
+pub struct StatisticValue {
+    pub raw: i64,
+    pub normalized: f64,
+}
+impl StatisticValue {
+    fn from_raw(raw: i64) -> StatisticValue {
+        StatisticValue { raw, normalized: f64::default() }
+    }
+}
+impl Ord for StatisticValue {
+    fn cmp(&self, other: &Self) -> Ordering {
+        (self.raw).cmp(&other.raw)
+    }
+}
+
+impl PartialOrd for StatisticValue {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Eq for StatisticValue {}
+
+#[derive(Clone, Copy)]
+pub struct StatisticValueRange {
+    pub max: StatisticValue,
+    pub min: StatisticValue,
+}
+impl StatisticValue {
+    pub fn normalize(&mut self, range: StatisticValueRange) {
+        self.normalized = (self.raw - range.min.raw) as f64 / (range.max.raw - range.min.raw) as f64
+    }
+}
+/// DocActivityMetrics stores key document activity features, which are used to recommend relevant documents to the user.
+/// Here's a walkthrough of the recommendation procedure: collect 1k most recent document events (write/read), use that activity to construct a DocActivtyMetrics struct for each document. Min-max normalizes the activity features, then rank the documents.
+/// latest_read_timestamp: the latest epoch timestamp that the user read a document
+/// latest_write_timestamp: the latest epoch timestamp that the user wrote a document
+/// the total number of times that a user reads a document
+/// the total number of times that a user wrote a document
+#[derive(Default, Copy, Clone)]
+pub struct DocActivityMetrics {
+    pub id: Uuid,
+    pub latest_read_timestamp: StatisticValue,
+    pub latest_write_timestamp: StatisticValue,
+    pub read_count: StatisticValue,
+    pub write_count: StatisticValue,
+}
+impl DocActivityMetrics {
+    pub fn rank(&self) -> i64 {
```

Comment:
rank usually indicates a position in a list (think valedictorian). score is what I think you're going for here (think gpa).